### ISSUE TITLE
Video stutter using VP8 over WHIP to Janus

### ISSUE THREAD

**Issue by lizardpeter** [CoSMoSoftware/OBS-studio-webrtc#278]
Hello,

I using the local network, Janus is reporting lots of "video_retransmissions" coming from OBS. I know you only offer support if the WHIP protocol is being used, so I am making sure to use that protocol. The playback is mostly fine, but using Chrome's "WebRTC internals" shows periods of packet loss (correlating to stutters in playback).

Thanks, 
Peter

**Comment by agouaillard-cosmo** [comment]
WHIP protocol being used is a necessary condition, but not sufficient. The support is still "best effort". We are all hands on deck for the v26 release. Let me check with lorenzo if his team has some cycles to help you.

**Comment by agouaillard-cosmo** [comment]
Hi @lminiero,
Do you have some cycle on your side for this?

**Comment by lminiero** [comment]
No, sorry, quite busy as we are. Anyway, I see Peter already wrote on our forum.

**Comment by agouaillard-cosmo** [comment]
Ah, sorry about the inconvenience, we were not aware of the duplicate post.

**Comment by agouaillard-cosmo** [comment]
@lminiero another question while at it, is your WHIP implementation public?

**Comment by lminiero** [comment]
Not at the moment, no. It wasn't particularly sophisticated, though: it was simply a POC that I used to validate the implementation. I think Peter is using a different WHIP signalling implementation provided by another Janus user.

**Comment by agouaillard-cosmo** [comment]
@lizardpeter How did you use WHIP with Janus exactly?

**Comment by lizardpeter** [comment]
I am using this implementation provided by Cameron Elliott: https://github.com/x186k/sfu186j.

**Comment by agouaillard-cosmo** [comment]
Thanks.

In any case, if your media is streaming through (however stuttering), it means the signalling / handshake, media path establishment, encryption all went through correctly.

Stuttering can come from bandwidth or CPU limitations, or bad network quality. To differentiate between the former two and the later, you could simply try sending a very low resolution like VGA.
- If you still experience the problem with low rez, your network is likely bad. 
- if the low resolution works well, start monitoring your bandwidth and your CPU usage and try with higher resolutions by small increment until you can reproduce the problem again.

HTH.

**Comment by lizardpeter** [comment]
Are you talking about CPU usage on the encoding side or on the Janus server ingestion side? Both seem to have fairly low usage overall during my tests. I will try to do it at a lower resolution later today, but from memory, I think I did experience stutters at the lower resolution. One thing that did seem to help a little was appending ?pkt_size=1200 to the url.

I can try to run some network tests, but this test is being run locally and both PCs are connected via cat6a (albeit at a gigabit link speed). Also, the external network is a gigabit fiber connection that (from my experience) has never had any packet loss.

**Comment by lminiero** [comment]
What Alex meant was that throughput in general can be impacted, especially since you mentioned very high bitrates that are never seen in webrtc (you talk of 15mbps, normal webrtc apps rarely go beyond 2). It also depends where you're experiencing this stuttering: if in a web page getting video from Janus, then either the browser can't cope with such high bitrate/fps/res, or we simply can't relay that many packets quickly enough.

**Comment by lizardpeter** [comment]
I am using the Janus VideoRoom demo page to view the stream. I will try again later today, but from my memory I do believe the stuttering was still occurring at around 720p 2 mbps. However, I have even tried at around 1080p 30 mbps 120 FPS and the playback itself is great at times while stuttering at others. Using the Janus recording feature, I looked at the recorded video and the stutter is not present in the recording. Apparently the RTP packets aren’t all arriving in order and within the specified interval (according to someone more knowledgeable than I am on the Google Group). The same stuttering also occurs on multiple web browsers at the same time, leading me to believe the issue is at the encoder or the connection from the encoder to Janus.

**Comment by lminiero** [comment]
Try the sendmmsg branch of Janus too (can't remember the PR number, I'm on my phone), as that's supposed to be more efficient in sending data.

**Comment by lizardpeter** [comment]
@lminiero Alright, I have tried to get the sendmmsg branch working, but I am getting a segmentation fault when running this version while the master version runs without a problem. Do you have any idea what the issue could be?

**Comment by lizardpeter** [comment]
Using the sendmmsg branch does appear to make things better, at least at lower bitrates and resolutions. However, I am trying to send this video out at at least 120 FPS, 1080p, and at as high of a bitrate as I can. 

@agouaillard-cosmo Thanks for the suggestion. Checking the CPU usage on the encoding PC, it is at around 25% usage even though NVENC is supposed to be being used (I can confirm that it is not being used because task manager is showing the GPU is not being used for encoding (maybe a bug?)). On the server PC, the CPU usage is at about 20%. In combination with Lorenzo's suggestion to use a different branch of Janus, it seems like videos at low bitrate and resolution are playing as they should. The issues seem to now occur after around 20000 kbps when using 720p and 120 FPS. Very few frames appear to be dropped at anything under those conditions. Are there any specific settings I should be using on the encoding side for maximum performance? Why is OBS not using NVENC?

**Comment by agouaillard-cosmo** [comment]
My gut feeling is that he available bandwidth is the problem.
OBS is using NVENC when sending RTMP.
OBS is using vanilla webrtc when sending webrtc, which only uses software codec.
As part of a more formal integration of webrtc support in webrtc, there is a design discussion ongoing with the OBS team to  unify codec usage.

**Comment by lizardpeter** [comment]
Bandwidth in terms of network bandwidth? The issues do seem to start to occur predictably at bitrates of around 20000 kbps. However, the network infrastructure here should be more than good enough to support much higher bitrates. I've seen YouTube use 100000 kbps or more when streaming some of the 8k 60 FPS videos.

Could it just be a limitation of WebRTC or maybe a setting on the encoder?

**Comment by agouaillard-cosmo** [comment]
I'm not aware of such limitation. However, I do not have an explicit test about this.
When #272 will be done, I will add a dedicated KITE test and we should try again.
ETA: a week or two.

**Comment by lizardpeter** [comment]
Awesome. Thank you!

**Comment by agouaillard-cosmo** [comment]
New deadline for #272 : March 6th

**Comment by lizardpeter** [comment]
Sounds good. I know this is not a typical use case for WebRTC, but this kind of technology - with its extremely low latency- would be exactly what I desire. I've tried HLS before, and it worked flawlessly with these very high bitrates and frame rates, but the latency was too high. Hopefully we can find a solution (whether the issue lies with OBS or Janus).

**Comment by agouaillard-cosmo** [comment]
Pushed back based on discussion about WHIP draft.
Depends now on #301

**Comment by lizardpeter** [comment]
Is there any update on this? I’m stuck using a much older version that has WHIP support.
